Source/Core/VideoBackends/D3D/DXPipeline.cpp

```diff
@@ -0,0 +1,264 @@
+  D3D::stateman->SetVertexShader(m_vertex_shader);
+  D3D::stateman->SetGeometryShader(m_geometry_shader);
+  D3D::stateman->SetPixelShader(m_pixel_shader);
+}
+
+void DXGraphicsPipeline::Unbind() const
+{
+  D3D::stateman->PopRasterizerState();
+  D3D::stateman->PopDepthState();
+  D3D::stateman->PopBlendState();
+}
+
+bool DXGraphicsPipeline::HasBinary() const
+{
+  return false;
+}
+
+AbstractShader::BinaryData DXGraphicsPipeline::GetBinary() const
+{
+  return {};
+}
+
+std::unique_ptr<AbstractGraphicsPipeline>
+Renderer::CreateGraphicsPipeline(const AbstractGraphicsPipelineConfig& config)
+{
+  ID3D11RasterizerState* rasterizer_state =
+      Renderer::GetStateCache().Get(config.rasterization_state);
+  ID3D11DepthStencilState* depth_state = Renderer::GetStateCache().Get(config.depth_state);
+  ID3D11BlendState* blend_state = Renderer::GetStateCache().Get(config.blending_state);
+  D3D11_PRIMITIVE_TOPOLOGY primitive_topology =
+      StateCache::GetPrimitiveTopology(config.rasterization_state.primitive);
+  if (!rasterizer_state || !depth_state || !blend_state)
+  {
+    SAFE_RELEASE(rasterizer_state);
+    SAFE_RELEASE(depth_state);
+    SAFE_RELEASE(blend_state);
+    return nullptr;
+  }
+
+  const DXShader* vertex_shader = static_cast<const DXShader*>(config.vertex_shader);
+  const DXShader* geometry_shader = static_cast<const DXShader*>(config.geometry_shader);
+  const DXShader* pixel_shader = static_cast<const DXShader*>(config.pixel_shader);
+  _assert_(vertex_shader != nullptr && pixel_shader != nullptr);
+
+  ID3D11InputLayout* input_layout =
+      const_cast<D3DVertexFormat*>(static_cast<const D3DVertexFormat*>(config.vertex_format))
+          ->GetInputLayout(vertex_shader->GetByteCode());
+
+  return std::make_unique<DXGraphicsPipeline>(
+      input_layout, vertex_shader->GetD3DVertexShader(),
+      geometry_shader ? geometry_shader->GetD3DGeometryShader() : nullptr,
+      pixel_shader->GetD3DPixelShader(), rasterizer_state, depth_state, blend_state,
+      primitive_topology);
+}
+
+bool Renderer::CreateUtilityResources()
+{
+  CD3D11_BUFFER_DESC vbo_desc(UTILITY_VBO_SIZE, D3D11_BIND_VERTEX_BUFFER, D3D11_USAGE_DYNAMIC,
+                              D3D11_CPU_ACCESS_WRITE);
+  HRESULT hr = D3D::device->CreateBuffer(&vbo_desc, nullptr, &m_utility_vertex_buffer);
+  CHECK(SUCCEEDED(hr), "Create utility VBO");
+  if (FAILED(hr))
+    return false;
+
+  CD3D11_BUFFER_DESC ubo_desc(UTILITY_UBO_SIZE, D3D11_BIND_CONSTANT_BUFFER, D3D11_USAGE_DYNAMIC,
+                              D3D11_CPU_ACCESS_WRITE);
+  hr = D3D::device->CreateBuffer(&ubo_desc, nullptr, &m_utility_uniform_buffer);
+  CHECK(SUCCEEDED(hr), "Create utility UBO");
+  if (FAILED(hr))
+    return false;
+
+  return true;
+}
+
+void Renderer::DestroyUtilityResources()
+{
+  SAFE_RELEASE(m_utility_vertex_buffer);
+  SAFE_RELEASE(m_utility_uniform_buffer);
+}
+
+void Renderer::DrawUtilityPipeline(const AbstractGraphicsPipeline* pipeline,
```

Comment:
Is there a reason all this `Renderer` stuff lives in the pipeline class?  I might be in the minority but I'm not  usually a fan of intermingling classes unless there's a good reason.